### PR TITLE
Add SQL syntax for GRANT/REVOKE on schema

### PR DESCRIPTION
**Pull request by lukasz-walkiewicz**

### CODE REVIEW

**Comment by lukasz-walkiewicz** [comment]
Resolves #4327

File: presto-main/src/main/java/io/prestosql/execution/GrantTask.java [hunk 1]
```diff
@@ -49,13 +54,50 @@ public String getName()
...
+            executeGrantOnTable(stateMachine.getSession(), statement, metadata, accessControl);
+        }
+        return immediateFuture(null);
+    }
+
+    private void executeGrantOnSchema(Session session, Grant statement, Metadata metadata, AccessControl accessControl)
+    {
+        CatalogSchemaName schemaName = createCatalogSchemaName(session, statement, Optional.of(statement.getName()));
+        Optional<CatalogName> catalogHandle = metadata.getCatalogHandle(session, schemaName.getCatalogName());
+        if (catalogHandle.isEmpty()) {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/execution/GrantTask.java:70]: `io.prestosql.metadata.Metadata#schemaExists`?

File: presto-main/src/main/java/io/prestosql/execution/RevokeTask.java [hunk 2]
```diff
@@ -49,13 +54,49 @@ public String getName()
...
+            executeRevokeOnTable(stateMachine.getSession(), statement, metadata, accessControl);
+        }
+        return immediateFuture(null);
+    }
+
+    private void executeRevokeOnSchema(Session session, Revoke statement, Metadata metadata, AccessControl accessControl)
+    {
+        CatalogSchemaName schemaName = createCatalogSchemaName(session, statement, Optional.of(statement.getName()));
+        Optional<CatalogName> catalogHandle = metadata.getCatalogHandle(session, schemaName.getCatalogName());
+        if (catalogHandle.isEmpty()) {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/execution/RevokeTask.java:70]: `io.prestosql.metadata.Metadata#schemaExists`?

File: presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4 [hunk 3]
```diff
@@ -87,12 +87,12 @@ statement
     | SET ROLE (ALL | NONE | role=identifier)                          #setRole
     | GRANT
         (privilege (',' privilege)* | ALL PRIVILEGES)
-        ON TABLE? qualifiedName TO grantee=principal
+        ON (SCHEMA | TABLE)? qualifiedName TO grantee=principal
```
Comment (kokosing) [inline presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4:90]:
```suggestion
        ON (SCHEMA | TABLE)? qualifiedName 
        TO grantee=principal
```

File: presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4 [hunk 4]
```diff
@@ -87,12 +87,12 @@ statement
...
         (WITH GRANT OPTION)?                                           #grant
     | REVOKE
         (GRANT OPTION FOR)?
         (privilege (',' privilege)* | ALL PRIVILEGES)
-        ON TABLE? qualifiedName FROM grantee=principal                 #revoke
+        ON (SCHEMA | TABLE)? qualifiedName FROM grantee=principal      #revoke
```
Comment (kokosing) [inline presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4:95]:
```suggestion
        ON (SCHEMA | TABLE)? qualifiedName 
        FROM grantee=principal      #revoke
```

File: presto-parser/src/main/java/io/prestosql/sql/tree/GrantOnType.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.tree;
+
+public enum GrantOnType
+{
+    SCHEMA,
+    TABLE
```
Comment (kokosing) [inline presto-parser/src/main/java/io/prestosql/sql/tree/GrantOnType.java:19]:
```suggestion
    SCHEMA,
    TABLE,
    /**/;
```

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParser.java [hunk 6]
```diff
@@ -1623,30 +1624,38 @@ public void testGrant()
...
         assertStatement("GRANT SELECT ON t TO ROLE PUBLIC WITH GRANT OPTION",
                 new Grant(
                         Optional.of(ImmutableList.of("SELECT")),
-                        false, QualifiedName.of("t"),
+                        Optional.empty(),
+                        QualifiedName.of("t"),
                         new PrincipalSpecification(PrincipalSpecification.Type.ROLE, new Identifier("PUBLIC")),
                         true));
-        assertStatement("GRANT ALL PRIVILEGES ON t TO USER u",
+        assertStatement("GRANT ALL PRIVILEGES ON TABLE t TO USER u",
```
Comment (kokosing) [inline presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParser.java:1638]: separate commit, you are changing not related test. As I understand this test case with present `TABLE` keyword in GRANT statement was missing, right?

**Comment by kokosing** [comment]
@martint Can you please double check the syntax?

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParser.java (hunk 6 above)
Comment (lukasz-walkiewicz) [reply presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParser.java:1638]: Correct

**Comment by kokosing** [comment]
> We do not have any connector that can support this

We are thinking about implementing this in Starburst.

> it's not even clear to how to add it existing connectors:

I am not sure I followed the previous discussion. I think we can revisit adding this to Hive SQL standard authorization.

> for other connector (say Postgres), how would this work? 

I didn't think hard about this, but to me it would simply translate to Postgres equivalent statements of GRANT and REVOKE to SCHEMA. 

> We would likely not want to configure a catalog with a Postgres superuser, but only a "normal" user. 

The best would be to use impersonation, credentials passthrough or extra credentials so user `bob` in Presto becomes `bob` in Postgres.

> Don't get me wrong

I don't. I appreciate when someone is questioning the idea before implementation. It could save hours of work.

File: presto-parser/src/main/java/io/prestosql/sql/SqlFormatter.java [hunk 7]
```diff
@@ -1366,10 +1367,11 @@ public Void visitRevoke(Revoke node, Integer indent)
             }
 
             builder.append(" ON ");
-            if (node.isTable()) {
-                builder.append("TABLE ");
+            if (node.getType().isPresent()) {
+                builder.append(node.getType().get().toString());
             }
-            builder.append(node.getTableName())
+            builder.append(" ");
```
Comment (kokosing) [inline presto-parser/src/main/java/io/prestosql/sql/SqlFormatter.java:1373]: don't append ` ` when `node.getType().isEmpty()`, same for grant

File: presto-main/src/main/java/io/prestosql/execution/RevokeTask.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/execution/RevokeTask.java:70]: ping

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/LegacyAccessControl.java [hunk 8]
```diff
@@ -247,6 +247,16 @@ public void checkCanSetCatalogSessionProperty(ConnectorSecurityContext context,
     {
     }
 
+    @Override
+    public void checkCanGrantSchemaPrivilege(ConnectorSecurityContext context, Privilege privilege, String schemaName, PrestoPrincipal grantee, boolean grantOption)
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/LegacyAccessControl.java:251]: I think this should accept a `Set<Privilege>` like `Metadata::grant*`. No change request, just for consideration for follow up because we would need to change check for tables as well.

File: presto-main/src/test/java/io/prestosql/connector/MockConnectorSchemaAccessControl.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.util.Arrays;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static io.prestosql.spi.security.AccessDeniedException.denyGrantSchemaPrivilege;
+import static io.prestosql.spi.security.AccessDeniedException.denyRevokeSchemaPrivilege;
+import static java.util.Objects.requireNonNull;
+
+class MockConnectorSchemaAccessControl
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/connector/MockConnectorSchemaAccessControl.java:29]: MockConnectorAccessControl? So it could be generic for other use cases?

File: presto-main/src/test/java/io/prestosql/connector/MockConnectorSchemaAccessControl.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        extends AllowAllAccessControl
+{
+    private static final String INFORMATION_SCHEMA = "information_schema";
+
+    private final SchemaGrants schemaGrants;
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/connector/MockConnectorSchemaAccessControl.java:34]: Make it optional. If it is empty let it work as allow all.

File: presto-main/src/test/java/io/prestosql/connector/SchemaGrants.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static io.prestosql.spi.security.PrincipalType.USER;
+import static java.util.Objects.requireNonNull;
+
+class SchemaGrants
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/connector/SchemaGrants.java:28]: public?

File: presto-tests/src/test/java/io/prestosql/execution/TestGrant.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    @Test
+    public void testDefaultGrants()
+    {
+        assertThat(assertions.query("SHOW SCHEMAS FROM local", admin).assertThat().matches("VALUES (VARCHAR 'information_schema'), (VARCHAR 'default')"));
+        assertThat(assertions.query("SHOW SCHEMAS FROM local", alice).assertThat().matches("VALUES (VARCHAR 'information_schema'), (VARCHAR 'default')"));
+        assertThat(assertions.query("SHOW SCHEMAS FROM local", bob).assertThat().matches("VALUES (VARCHAR 'information_schema')"));
+    }
+
+    @Test
+    public void testGrantOnSchema()
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/execution/TestGrant.java:95]: Add tests for:
GRANT/REVOKE ALL
GRANT/REVOKE GRANT OPTION

File: presto-main/src/test/java/io/prestosql/connector/SchemaGrants.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+{
+    private final Map<SchemaGrant, SchemaGrant> grants = new HashMap<>(); // Using map to be able to manipulate an element on insert
```
[the remark this answers is not in the document]
Comment (lukasz-walkiewicz) [reply presto-main/src/test/java/io/prestosql/connector/SchemaGrants.java]: Set doesn't have a method to get an existing entry. I've left a comment.

[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/test/java/io/prestosql/connector/SchemaGrants.java]: This is very misleading then. And also it looks like you had to hack hashCode/equals methods.

How about having a real map? `Map<String /* schema */, Grant>` or `Map<UserAndSchema, Grant>`?

**Comment by martint** [comment]
Syntax-wise, this is not in the SQL spec, but it's consistent with it:

```
<grant privilege statement> ::=
GRANT <privileges> TO <grantee> [ { <comma> <grantee> }... ]
      [ WITH HIERARCHY OPTION ]
      [ WITH GRANT OPTION ]
      [ GRANTED BY <grantor> ]

<privileges> ::=
  <object privileges> ON <object name>

<object name> ::= 
      [ TABLE ] <table name>
    | DOMAIN <domain name>
    | COLLATION <collation name>
    | CHARACTER SET <character set name> 
    | TRANSLATION <transliteration name> 
    | TYPE <schema-resolved user-defined 
    | SEQUENCE <sequence generator name> 
    | <specific routine designator>
````

What I think needs to be discussed are the semantics of this operation, given that it introduces a behavior on hierarchical objects that isn't contemplated by the spec. I.e., does revoking a privilege on a schema imply a revocation of privileges over all tables in the schema? Does performing a SELECT require checking permissions both on the table and the schema? Etc...

For reference, the spec describes SELECT permissions as applying to columns, not tables. Doing a `GRANT SELECT ON TABLE t` is a way to say "grant SELECT on current and future columns of the table". `GRANT SELECT(x) ON TABLE t` means "grant SELECT only on column x of table t".

```
2) SELECT (<column name>) specifies the SELECT privilege on the indicated column and implies one or more column privilege descriptor.
[...]
4) SELECT with neither <privilege column list> nor <privilege method list> specifies the SELECT privilege on all columns of T including any columns subsequently added to T and implies a table privilege descriptor and one or more column privilege descriptors. If T is a table of a structured type TY, then SELECT also specifies the SELECT privilege on all methods of the type TY, including any methods subsequently added to the type TY, and implies one or more table/method privilege descriptors.
```

It records the privilege in a table privilege descriptor for the purpose of applying the privilege to columns that are added afterwards, but only the column privilege descriptors are consulted when querying. From section "6.7 <column reference>":

```
ii) Otherwise, the current privileges shall include SELECT on the column referenced by CR.
```

**Comment by kokosing** [comment]
> WITH HIERARCHY OPTION 

What does it mean?

>  does revoking a privilege on a schema imply a revocation of privileges over all tables in the schema

Schema privilege is to me independent to other privileges. So revoking should not affect any other privileges that user currently has.

So
```
GRANT SELECT on s.t TO bob;
 -- bob can select from s.t
GRANT SELECT SCHEMA s TO bob;
 -- bob can select from s.t
REVOKE SELECT SCHEMA s FROM bob;
 -- bob can select from s.t
REVOKE SELECT on s.t FROM bob;
-- bob can no longer select from s.t
```

> Doing a GRANT SELECT ON TABLE t is a way to say "grant SELECT on current and future columns of the table". 

So doing `GRANT SELECT ON SCHEMA s` is a way to say "grant SELECT on current and future columns of all current and future tables of SCHEMA s" 

> Does performing a SELECT require checking permissions both on the table and the schema?

No, because the above (GRANT SELECT ON SCHEMA translates to grant SELECT on columns).

> What I think needs to be discussed are the semantics of this operation,

One more point. We try to follow data source semantics when writing connectors. So Hive connector should behave like Hive. So should we do the same thing here for access control. For example should allow to have different semantics when `hive.security=sql-standard` or `hive.security=other-authorization-polices-service` so we follow the behavior of `sql-standard` and `other-authorization-polices-service` in hive?

File: presto-docs/src/main/sphinx/sql/grant.rst [hunk 14]
```diff
@@ -8,7 +8,7 @@ Synopsis
 .. code-block:: none
 
     GRANT ( privilege [, ...] | ( ALL PRIVILEGES ) )
-    ON [ TABLE ] table_name TO ( user | USER user | ROLE role )
+    ON ( table_name | TABLE table_name | SCHEMA schema_name) TO ( user | USER user | ROLE role )
```
Comment (kokosing) [inline presto-docs/src/main/sphinx/sql/grant.rst:11]:
```suggestion
    ON ( table_name | TABLE table_name | SCHEMA schema_name) 
    TO ( user | USER user | ROLE role )
```

File: presto-docs/src/main/sphinx/sql/revoke.rst [hunk 15]
```diff
@@ -9,7 +9,7 @@ Synopsis
 
     REVOKE [ GRANT OPTION FOR ]
     ( privilege [, ...] | ALL PRIVILEGES )
-    ON [ TABLE ] table_name FROM ( user | USER user | ROLE role )
+    ON ( table_name | TABLE table_name | SCHEMA schema_name ) FROM ( user | USER user | ROLE role )
```
Comment (kokosing) [inline presto-docs/src/main/sphinx/sql/revoke.rst:12]:
```suggestion
    ON ( table_name | TABLE table_name | SCHEMA schema_name ) 
    FROM ( user | USER user | ROLE role )
```

File: presto-docs/src/main/sphinx/sql/grant.rst [hunk 16]
```diff
@@ -8,7 +8,7 @@ Synopsis
...
     [ WITH GRANT OPTION ]
 
 Description
```
Comment (kokosing) [inline presto-docs/src/main/sphinx/sql/grant.rst:14]: Mention the semantic of grant privileges to table (in separate commit) and then for schema

File: presto-docs/src/main/sphinx/sql/revoke.rst [hunk 17]
```diff
@@ -9,7 +9,7 @@ Synopsis
...
 
 Description
```
Comment (kokosing) [inline presto-docs/src/main/sphinx/sql/revoke.rst:14]: Mention the semantic of revoke privileges to table (in separate commit) and then for schema

File: presto-tests/src/test/java/io/prestosql/common/Randoms.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+ */
+package io.prestosql.common;
+
+import java.util.concurrent.ThreadLocalRandom;
+
+public final class Randoms
+{
+    private Randoms() {}
+
+    public static String randomUsername()
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/common/Randoms.java:22]: `io.prestosql.testing.sql.TestTable#randomTableSuffix`?

File: presto-main/src/main/java/io/prestosql/execution/RevokeTask.java [hunk 19]
```diff
@@ -49,13 +54,52 @@ public String getName()
...
+        return immediateFuture(null);
+    }
+
+    private void executeRevokeOnSchema(Session session, Revoke statement, Metadata metadata, AccessControl accessControl)
+    {
+        CatalogSchemaName schemaName = createCatalogSchemaName(session, statement, Optional.of(statement.getName()));
+
+        if (!metadata.catalogExists(session, schemaName.getCatalogName())) {
+            throw semanticException(CATALOG_NOT_FOUND, statement, "Catalog '%s' does not exist", schemaName.getCatalogName());
+        }
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/execution/RevokeTask.java:72]:
```suggestion
```

I think this is not needed, the below check is good enough.

File: presto-main/src/main/java/io/prestosql/execution/GrantTask.java [hunk 20]
```diff
@@ -49,13 +54,53 @@ public String getName()
...
+        return immediateFuture(null);
+    }
+
+    private void executeGrantOnSchema(Session session, Grant statement, Metadata metadata, AccessControl accessControl)
+    {
+        CatalogSchemaName schemaName = createCatalogSchemaName(session, statement, Optional.of(statement.getName()));
+
+        if (!metadata.catalogExists(session, schemaName.getCatalogName())) {
+            throw semanticException(CATALOG_NOT_FOUND, statement, "Catalog '%s' does not exist", schemaName.getCatalogName());
+        }
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/execution/GrantTask.java:72]:
```suggestion
```
I think this is not needed, the below check is good enough.

File: presto-main/src/test/java/io/prestosql/connector/MockConnectorAccessControl.java [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    public Set<String> filterSchemas(ConnectorSecurityContext context, Set<String> schemaNames)
+    {
+        return schemaNames.stream()
+                .filter(schema ->
+                        schema.equalsIgnoreCase(INFORMATION_SCHEMA) ||
+                                Arrays.stream(Privilege.values())
+                                        .anyMatch(privilege -> schemaGrants.isAllowed(context.getIdentity().getUser(), schema, privilege)))
+                .collect(Collectors.toSet());
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/connector/MockConnectorAccessControl.java:49]: toImmutableSet

File: presto-main/src/test/java/io/prestosql/connector/MockConnectorAccessControl.java [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Override
+    public Set<String> filterSchemas(ConnectorSecurityContext context, Set<String> schemaNames)
+    {
+        return schemaNames.stream()
+                .filter(schema ->
+                        schema.equalsIgnoreCase(INFORMATION_SCHEMA) ||
+                                Arrays.stream(Privilege.values())
+                                        .anyMatch(privilege -> schemaGrants.isAllowed(context.getIdentity().getUser(), schema, privilege)))
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/connector/MockConnectorAccessControl.java:48]: you might want to extract this condition as method

File: presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4 [hunk 23]
```diff
@@ -87,12 +87,14 @@ statement
...
+        TO grantee=principal
         (WITH GRANT OPTION)?                                           #grant
     | REVOKE
         (GRANT OPTION FOR)?
         (privilege (',' privilege)* | ALL PRIVILEGES)
-        ON TABLE? qualifiedName FROM grantee=principal                 #revoke
+        ON (SCHEMA | TABLE)? qualifiedName
+        FROM grantee=principal                                         #revoke
     | SHOW GRANTS
         (ON TABLE? qualifiedName)?                                     #showGrants
```
Comment (kokosing) [inline presto-parser/src/main/antlr4/io/prestosql/sql/parser/SqlBase.g4:99]: In docs  you mentioned that `SHOW GRANTS ON SCHEMA` is supported, please undo the change in docs

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedAccessControl.java [hunk 24]
```diff
@@ -306,6 +308,22 @@ public void checkCanSetCatalogSessionProperty(ConnectorSecurityContext context,
         }
     }
 
+    @Override
+    public void checkCanGrantSchemaPrivilege(ConnectorSecurityContext context, Privilege privilege, String schemaName, PrestoPrincipal grantee, boolean grantOption)
+    {
+        if (!isSchemaOwner(context, schemaName)) {
```
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedAccessControl.java:314]: Please add tests for that

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java [hunk 25]
```diff
@@ -660,6 +662,28 @@ public void checkCanSetCatalogSessionProperty(SystemSecurityContext context, Str
     {
     }
 
+    @Override
+    public void checkCanGrantSchemaPrivilege(SystemSecurityContext context, Privilege privilege, CatalogSchemaName schema, PrestoPrincipal grantee, boolean grantOption)
```
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedSystemAccessControl.java:666]: Please add tests for that

File: presto-tests/src/test/java/io/prestosql/execution/TestGrant.java [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+
+    @AfterClass(alwaysRun = true)
+    public void teardown()
+    {
+        assertions.close();
+        assertions = null;
+    }
+
+    @Test
+    public void testDefaultGrants()
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/execution/TestGrant.java:96]: While you are it, can you please also add few tests for GRANT/REVOKE ON TABLE?

File: presto-tests/src/test/java/io/prestosql/execution/TestGrant.java [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+import java.util.EnumSet;
+
+import static io.prestosql.common.Randoms.randomUsername;
+import static io.prestosql.spi.security.PrincipalType.USER;
+import static io.prestosql.testing.TestingSession.testSessionBuilder;
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestGrant
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/execution/TestGrant.java:42]: Please also add tests for SHOW GRANTS and `information_schema.table_privileges` while using GRANT/REVOKE ON SCHEMA

File: presto-tests/src/test/java/io/prestosql/execution/TestRevoke.java [hunk 28]
```diff
@@ -0,0 +1,134 @@
...
+import java.util.EnumSet;
+
+import static io.prestosql.common.Randoms.randomUsername;
+import static io.prestosql.spi.security.PrincipalType.USER;
+import static io.prestosql.testing.TestingSession.testSessionBuilder;
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestRevoke
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/execution/TestRevoke.java:42]: I would merge `TestGrant` with `TestRevoke`.  :)

**Review by kokosing — COMMENTED** [review]
This looks decent. Nice! Just few comments otherwise looks good.

@martint Please take a look

File: presto-tests/src/test/java/io/prestosql/execution/TestGrantOnSchema.java [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+    @DataProvider(name = "privileges")
+    public static Object[][] privileges()
+    {
+        return new Object[][] {
+                {"SELECT"},
+                {"ALL PRIVILEGES"}
+        };
+    }
+
+    private static Session sessionOf(String username)
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/execution/TestGrantOnSchema.java:143]: just `session`?

File: presto-tests/src/test/java/io/prestosql/execution/TestRevokeOnSchema.java [hunk 30]
```diff
@@ -0,0 +1,132 @@
...
+    @DataProvider(name = "privileges")
+    public static Object[][] privileges()
+    {
+        return new Object[][] {
+                {"SELECT"},
+                {"ALL PRIVILEGES"}
+        };
+    }
+
+    private static Session sessionOf(String username)
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/execution/TestRevokeOnSchema.java:124]: session?

**Review by kokosing — APPROVED** [review]
@martint Would you like to take a look?

**Comment by kokosing** [comment]
Merged, thanks!
